plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/ReusingConnectionFactory.java

```diff
@@ -0,0 +1,205 @@
+        }
+        return delegate.openConnection(session);
+    }
+
+    @Override
+    public void beginQuery(ConnectorSession session) {}
+
+    @Override
+    public void cleanupQuery(ConnectorSession session)
+    {
+        Connection connection = connections.remove(session.getQueryId());
+        if (connection != null) {
+            try {
+                connection.close();
+            }
+            catch (SQLException e) {
+                throw new TrinoException(JDBC_ERROR, e);
+            }
+        }
+    }
+
+    @Override
+    public void close()
+            throws SQLException
+    {
+        for (Connection connection : connections.values()) {
+            connection.close();
+        }
+        connections.clear();
+        delegate.close();
+    }
+
+    private final class CachedConnection
+            extends ForwardingConnection
+    {
+        private final String queryId;
+        private final Connection delegate;
+        private boolean closed;
+        private boolean dirty;
+
+        private CachedConnection(String queryId, Connection delegate)
+        {
+            this.queryId = requireNonNull(queryId, "queryId is null");
+            this.delegate = requireNonNull(delegate, "delegate is null");
+        }
+
+        @Override
+        protected Connection getDelegate()
+        {
+            checkState(!closed, "Connection is already closed");
+            return delegate;
+        }
+
+        @Override
+        public <T> T unwrap(Class<T> iface)
+                throws SQLException
+        {
+            dirty = true;
+            return super.unwrap(iface);
+        }
+
+        @Override
+        public void setReadOnly(boolean readOnly)
+                throws SQLException
+        {
+            dirty = true;
+            super.setReadOnly(readOnly);
+        }
+
+        @Override
+        public void setCatalog(String catalog)
+                throws SQLException
+        {
+            dirty = true;
+            super.setCatalog(catalog);
+        }
+
+        @Override
+        public void setSchema(String schema)
+                throws SQLException
+        {
+            dirty = true;
+            super.setSchema(schema);
+        }
+
+        @Override
+        public void setTransactionIsolation(int level)
+                throws SQLException
+        {
+            dirty = true;
+            super.setTransactionIsolation(level);
+        }
+
+        @Override
+        public void setAutoCommit(boolean autoCommit)
+                throws SQLException
+        {
+            dirty = true;
+            super.setAutoCommit(autoCommit);
+        }
+
+        @Override
+        public void close()
+                throws SQLException
+        {
+            if (closed) {
+                return;
+            }
+            closed = true;
+            if (dirty) {
+                delegate.close();
+            }
+            else {
+                connections.put(queryId, delegate);
```

Comment:
What if two connections were acquired for a query? For example in a recursive handling of `information_schema.columns` or anything like that. 
We would overwrite entry in the cache. Will both connections get closed?

Reply:
What if something holds to a connection and the connection gets closed during `io.trino.plugin.jdbc.QueryEventListener#cleanupQuery`?

Thus, we can do `connections.put(queryId, delegate)`
after `connections.remove(session.getQueryId())` has been called, and thus leak a connection.
To prevent this from happening, you can wrap cache value (the connection) with "a connection holder" that would be stateful: `{ DISCONNECTED, CONNECTED, LEASED, CLOSED }`. 

This also remove explicit remove/replace from the cache, so `.removalListener` would be simpler and more obviously correct.